### PR TITLE
Add type hints

### LINKED ISSUE #745: Add typing
Hey guys,

thanks for creating this package. Short question: Are you open to add type hints (typing) to this package?
Basically. I would like mypy to check the usage and currently this is not possible.

I'm also open to create a PR.

### PR DESCRIPTION
**Pull request by kasium**
Closes #745

### CODE REVIEW

**Comment by kasium** [comment]
@lafrech I added as many type hints as possible. Can you please have a look?

File: setup.cfg [hunk 1]
```diff
@@ -6,3 +6,14 @@ ignore = E203, E266, E501, W503
...
+[mypy]
+ignore_missing_imports = true
+warn_unused_configs = True
+no_implicit_optional = True
+warn_redundant_casts = True
+warn_unused_ignores = True
+show_error_context = True
+show_column_numbers = True
+show_error_codes = True
+pretty = True
```
Comment (lafrech) [inline setup.cfg]: I'd like to keep this harmonized with marshmallow/webargs/... as much as possible.

https://github.com/marshmallow-code/marshmallow/blob/dev/setup.cfg

```
[mypy]
ignore_missing_imports = true
warn_unreachable = true
warn_unused_ignores = true
warn_redundant_casts = true
no_implicit_optional = true
```

File: src/apispec/yaml_utils.py [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
 import yaml
+import typing
 
 from apispec.utils import trim_docstring, dedent
 
 
-def dict_to_yaml(dic, yaml_dump_kwargs=None):
+def dict_to_yaml(
+    dic: typing.Any, yaml_dump_kwargs: typing.Optional[typing.Any] = None
+) -> str:
```
Comment (lafrech) [inline src/apispec/yaml_utils.py]: Isn't `dic` of type `dict`?

**Review by lafrech — COMMENTED** [review]
Thank you very much already for the contribution.

I didn't review thoroughly already, but this looks good.

I think we should use 

```
from __future__ import annotations
```

as in https://github.com/marshmallow-code/marshmallow/pull/1932.

**Comment by kasium** [comment]
> Thank you very much already for the contribution.
> 
> I didn't review thoroughly already, but this looks good.
> 
> I think we should use
> 
> ```
> from __future__ import annotations
> ```
> 
> as in [marshmallow-code/marshmallow#1932](https://github.com/marshmallow-code/marshmallow/pull/1932).

good idea. I changed the hints to that system

**Comment by kasium** [comment]
I got this weird CI error: `No hosted parallelism has been purchased or granted. To request a free parallelism grant, please fill out the following form https://aka.ms/azpipelines-parallelism-request`

**Comment by lafrech** [comment]
Strange. We're about to migrate from Azure, anyway. To GH actions + pre-commit.ci.

Please give me a little time to check this. Thanks again.

**Review by lafrech — COMMENTED** [review]
Looks pretty good.

Do you think you could squash all your commits and push-force?

This would provide a nicer commit history and it would also trigger a new CI run. Not sure what failed with Azure, let's see how it goes this time.

Thanks.

**Comment by kasium** [comment]
@lafrech the CI still doesn't work. I tried to squash everything but merge commits and rebases are not nice.
However, if you merge the PR, you can normally also select squash.

**Comment by lafrech** [comment]
Indeed, I read a bit more about the CI error message. I don't want to struggle with this while about to change CI anyway. I shall move to Github actions sooner than I expected.

**Comment by lafrech** [comment]
@kasium sorry about the delay. CI issues should now be fixed.

Do you think you could rebase to let the tests run before we merge?

Thanks !

**Comment by kasium** [comment]
@lafrech done

**Comment by lafrech** [comment]
Thank you so much!

**Pull request merged**